src/dienstplan/db.clj

```diff
@@ -182,52 +199,48 @@
        first
        :user))
 
+(defn- update-users
+  [conn users ts]
+  (let [rota_id (first users)
+        users-updated
+        (reduce
+          +
+          (map
+            (fn [user]
+              (first
+                (jdbc/update!
+                  conn :mention
+                  {:duty (:duty user)}
+                  ["id = ?" (:id user)])))
+            users))
+        _
+        (when rota_id
+          (jdbc/update!
+            conn :rota
+            {:updated_on ts}
+            ["id = ?" (:rota_id rota_id)]))]
+    users-updated))
+
 (defn rotate-duty!
   [channel rotation ts]
   (jdbc/with-db-transaction [conn db]
     (let
-     [users
-      (into
-       []
-       (jdbc/query
-        conn
-        ["SELECT
-              m.id,
-              m.rota_id,
-              m.name AS user,
-              m.duty
-            FROM rota AS r
-            JOIN mention AS m ON r.id = m.rota_id
-            WHERE
-              1 = 1
-              AND r.channel = ?
-              AND r.name = ?
-            ORDER BY r.id ASC
-            FOR UPDATE"
-         channel rotation]))
+     [users (rota-get conn channel rotation)
       users-count (count users)
       rotated (rotate-users users)
       prev-duty (get-duty-user-name users)
       current-duty (get-duty-user-name rotated)
-      rota_id (first users)
-      users-updated
-      (reduce
-       +
-       (map
-        (fn [user]
-          (first
-           (jdbc/update!
-            conn :mention
-            {:duty (:duty user)}
-            ["id = ?" (:id user)])))
-        rotated))
-      _
-      (when rota_id
-        (jdbc/update!
-         conn :rota
-         {:updated_on ts}
-         ["id = ?" (:rota_id rota_id)]))]
+      users-updated (update-users conn rotated ts)]
       {:users-count users-count
        :users-updated users-updated
        :prev-duty prev-duty
        :current-duty current-duty})))
+
+(defn assign!
+  [channel rotation name ts]
+  (jdbc/with-db-transaction [conn db]
+    (let
+      [users (rota-get conn channel rotation)
+       assigned (assign-user users name)
+       _ (update-users conn assigned ts)]
```

Comment:
This line will fail miserably if `assigned` returns `:user-not-found`:

```
(first (assign-user [] nil))
```

We must check for `:user-not-found` and invoke `update-users` only if a user found.

Because it can pop up in a variety of scenarios:

```
# omit user mention at all
@diesntplan assign my-rota

# use text instead of mention for regexp to fail
@diesntplan assign my-rota text

# mention user not included in the rota
@dienstplan assing my-rota @user-not-from-rota
```

Reply:
You're right! Fixed